### ISSUE TITLE
Ordering-dependent issue running queries

### ISSUE THREAD

**Issue by loganmhb** [loganmhb/logos#32]
These should be equivalent:
```
> find ?e ?a ?v where (?e ?a ?v) (?a db:ident ?attr)
+---+---+-----------------------------------+
| e | a |                 v                 |
+---+---+-----------------------------------+
| 0 | 2 | 2017-09-09 00:47:55.789457110 UTC |
| 1 | 1 |             db:ident              |
| 1 | 3 |           db:type:ident           |
| 2 | 1 |           db:txInstant            |
| 3 | 1 |           db:valueType            |
| 3 | 3 |           db:type:ident           |
| 4 | 1 |           db:type:ident           |
| 5 | 1 |          db:type:string           |
| 6 | 1 |         db:type:timestamp         |
| 7 | 1 |          db:type:entity           |
+---+---+-----------------------------------+


> find ?e ?a ?v where (?a db:ident ?attr) (?e ?a ?v)
ERROR: Error("type mismatch")
```

The issue seems to be with `Clause.substitute()`, which can't substitute correctly into the attribute position because attributes are stored in the DB as entities but represented in a `Clause` by a string.

**Comment by loganmhb** [comment]
This bug is a result of semi-pervasive ambiguity between whether attributes are represented by their names or by their entities -- it would be good to nail down exactly where they are converted from one to the other and make that consistent. The main question is whether access to the DB ident map should be a prerequisite for parsing a query.

**Comment by loganmhb** [comment]
This will be resolved by #33 .
